### ISSUE TITLE
MultiClass AUC

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr#3]
See wether we can create an AUC measure for more than 2 classes in mlr.

Here is a hint, sent by Markus by mail.

```r
 learner <- makeLearner('classif.lda', predict.type="prob")
 task <- makeClassifTask(data=iris, target="Species")
 mod <- train(learner=learner, task=task)
 pred.obj <- predict(mod, newdata=iris)
 library(HandTill2001)
 predicted <- as.matrix(pred.obj$data[,paste("prob.", levels(pred.obj$data$response), sep="")])
 colnames(predicted)<-levels(pred.obj$data$response)
 auc(multcap(response=pred.obj$data$response, predicted=predicted))
```

1) Investigate

2) Add Measure, doc. and test it

3) Briefly describe in tutorial / ROC part.

**Comment by PhilippPro** [comment]
I think there is something wrong in the multiclass.auc function. As input there is only the response and the probability, but not the truth. A prediction cannot be evaluated without the true values. Probably the term `response` in the pROC package and the poor documentation is missleading. 
In my opinion `pred$data$response` has to be replaced with `pred$data$truth`.

Also the multiclass.auc and auc should coincide in the two class case. We could transform the probabilities like in #858 or just apply auc. 

Maybe it would also be better to implement the auc stuff by ourselves, maybe I will do it. In this paper there are described several different auc measures for the multiclass case:
https://www.math.ucdavis.edu/~saito/data/roc/ferri-class-perf-metrics.pdf

**Comment by jakob-r** [comment]
I think you are pretty right. We should fix this asap. But the documentation does not enlighten me so much what the correct input should be. But I can see that the current state can not work.

**Comment by catastrophic-failure** [comment]
An example using `caTools`

```splus
library(mlr)

lrn = makeLearner("classif.randomForest", predict.type = "prob")
model = train(lrn, iris.task)
pred = predict(model, iris.task)

library(caTools)
colAUC(getPredictionProbabilities(pred), pred$data$truth)
```

> 
> > colAUC(getPredictionProbabilities(pred), pred$data$truth)
>                                     setosa  versicolor  virginica
> setosa vs. versicolor    1.00     1.0000       0.84
> setosa vs. virginica      1.00      0.7336      1.00
> versicolor vs. virginica  0.54     1.0000       1.00

Each column of probability estimates is a 1-vs-rest binary classifier. Then we can group the outputted AUCs knowing the labels.

**Comment by berndbischl** [comment]
@Coorsaa Stefan, please also take look at this.

**Comment by catastrophic-failure** [comment]
I made a small code to select all the possible pairwise combinations to evaluate Hand & Till AUC. I proof tested it on the iris data set with some added dummy classes.

```splus
m = colAUC(getPredictionProbabilities(pred), pred$data$truth)

c = t(combn(1:nlevels(pred$data$truth), 2, simplify = T))

m[cbind(1:nrow(c),c[,2])]
```

**Comment by PhilippPro** [comment]
This looks very good! I also tested the colAUC function against the normal auc function in pROC and they seem to provide the same values (if the input values in auc are restricted to certain kinds of levels/probabilities). 
I would suggest to write our own function of colAUC in mlr package (in fact it is open-source) instead of having a new dependency. What do you think about that? With the colAUC function it will also be easy to calculate the other multiclass AUC measures.

**Comment by PhilippPro** [comment]
Here the other measures with the help of colAUC:

https://github.com/PhilippPro/MulticlassAUC/blob/master/MulticlassAUC.R

After having a feedback about putting a new function colAUC in mlr or a new dependency, I can implement them in mlr and make a pull request.

**Comment by PhilippPro** [comment]
Ok, after some combinatorical thinking I think your code (@catastrophical-failure), was not completely right for more than 3 classes. I think a possible solution would be the following:

`m = colAUC(probabilities, pred$data$truth)
c = c(combn(1:nlevels(pred$data$truth), 2))
mean(m[cbind(rep(1:nrow(m), each = 2), c)])`

Do you agree?

**Comment by catastrophic-failure** [comment]
I tested it (mine and yours) on 5 classes; yours also picks every possible pairwise combination, but it also takes the pairs AUC(j,k) and AUC(k,j), the result is the same, but mine excludes the identical pairwise measures.

**Comment by catastrophic-failure** [comment]
Yeah, you are right. I was under the impression that pairwise AUCs would be equal among the pair, but that isn't truth.

**Comment by jakob-r** [comment]
PR #910

**Comment by berndbischl** [comment]
fixed in #960
